lib/p2p/Peer.ts

```diff
@@ -1,20 +1,251 @@
+  }
+
+  private _init = () => {
+    this.parser.on('packet', async (packet) => {
+
+      // handle packet in the Peer level here, if necessary.
+
+      this.emit('packet', packet);
+    });
+
+    this.parser.on('error', (err: ParserError) => {
+      if (this.destroyed) {
+        return;
+      }
+
+      switch (err.type) {
+        case ParserErrorType.UNPARSABLE_MESSAGE: {
+          this.logger.warn(`Unparsable peer message: ${err.payload}`);
+          this.error(err);
+          this.increaseBan(10);
+          break;
+        }
+        case ParserErrorType.UNKNOWN_PACKET_TYPE: {
+          this.logger.warn(`Unknown peer message type: ${err.payload}`);
+          this.error(err);
+          this.increaseBan(20);
+        }
+      }
+    });
+  }
+
+  private increaseBan = (score): boolean => {
+    this.banScore += score;
+
+    if (this.banScore >= 100) { // TODO: make configurable
+      this.logger.debug(`Ban threshold exceeded (${this.id})`);
+      this.emit('ban');
+      return true;
+    }
+
+    return false;
+  }
 
-Peer.prototype.toString = function toString() {
-  const { host, address, port } = this;
-  return !host || address === host
-      ? `${address}:${port}`
-      : `${host}[${address}]:${port}`;
-};
+  private initConnection = (): Promise<void> => {
+    assert(this.socket);
+
+    if (this.connected) {
+      assert(this.direction === ConnectionDirection.INBOUND);
+      this.logger.debug(this.statusString);
+      return Promise.resolve();
+    }
+
+    return new Promise((resolve, reject) => {
+      const cleanup = () => {
+        if (this.connectTimeout) {
+          clearTimeout(this.connectTimeout);
+          this.connectTimeout = undefined;
+        }
+        this.socket.removeListener('error', onError);
+      };
+
+      const onError = (err) => {
+        cleanup();
+        reject(err);
+      };
+
+      this.socket.once('connect', () => {
+        this.connectTime = Date.now();
+        this.connected = true;
+        this.logger.debug(this.statusString);
+        this.emit('connect');
+
+        cleanup();
+        resolve();
+      });
+
+      this.socket.once('error', onError);
+
+      this.connectTimeout = setTimeout(() => {
+        this.connectTimeout = undefined;
+        cleanup();
+        reject(new Error('Connection timed out.'));
+      }, 10000);
+    });
+  }
+
+  private connect = (address: NetAddress): void => {
+    assert(!this.socket);
+
+    const socket = net.connect(address.port, address.host);
+
+    this.address = address;
+    this.direction = ConnectionDirection.OUTBOUND;
+    this.connected = false;
+
+    this.bind(socket);
+  }
+
+  private accept = (socket: Socket): void => {
+    assert(!this.socket);
+
+    this.address = NetAddress.fromSocket(socket);
+    this.direction = ConnectionDirection.INBOUND;
+    this.connected = true;
+
+    this.bind(socket);
+  }
+
+  private bind = (socket: Socket) => {
```

Comment:
I don’t think that utilizing a setter would be the best approach here, although it can technically work. might be too confusing, and not explicit enough.

I think we should call this `bindSocket`, since it’s not really about initializing the socket.

Reply:
My thinking is that we will always be binding the event handlers whenever we set the socket, rolling them into a single piece of logic makes sense. But `bindSocket` also makes sense, so I'll defer to what you think is best.

Reply:
You're right, but since it doesn't meant to be used besides in the object instantiation, I would rather keep it separated.